proctor-common/src/main/java/com/indeed/proctor/common/TestChooser.java

```diff
@@ -25,15 +29,67 @@
     @Nonnull
     String getTestName();
 
+    /**
+     * Do not directly call this outside this interface.
+     * We should call {@link #choose(Object, Map, Map, ForceGroupsOptions)}, instead.
+     */
+    @Nonnull
+    TestChooser.Result chooseInternal(
+            @Nullable IdentifierType identifier,
+            @Nonnull Map<String, Object> values,
+            @Nonnull Map<String, TestBucket> testGroups
+    );
+
     @Nonnull
-    TestChooser.Result choose(@Nullable IdentifierType identifier, @Nonnull Map<String, Object> values, @Nonnull Map<String, TestBucket> testGroups);
+    default TestChooser.Result choose(
+            @Nullable IdentifierType identifier,
+            @Nonnull Map<String, Object> values,
+            @Nonnull Map<String, TestBucket> testGroups,
+            @Nonnull ForceGroupsOptions forceGroupsOptions
+    ) {
+        final String testName = getTestName();
+
+        final Optional<Integer> forceGroupBucket = forceGroupsOptions.getForcedBucketValue(testName);
+        if (forceGroupBucket.isPresent()) {
+            final TestBucket forcedTestBucket = getTestBucket(forceGroupBucket.get());
+            if (forcedTestBucket != null) {
+                // use a forced bucket, skip choosing an allocation
+                return new Result(forcedTestBucket, null);
+            }
+        }
+
+        if (forceGroupsOptions.getDefaultMode().equals(ForceGroupsDefaultMode.FALLBACK)) {
+            // skip choosing a test bucket and an allocation
+            return Result.EMPTY;
+        }
+
+        final TestChooser.Result result = chooseInternal(identifier, values, testGroups);
+
+        if (forceGroupsOptions.getDefaultMode().equals(ForceGroupsDefaultMode.MIN_ACTIVE)) {
+            // replace the bucket with the minimum active bucket in the resolved allocation.
+            return Optional.ofNullable(result.getAllocation())
+                    .map(Allocation::getRanges)
+                    .map(Collection::stream)
+                    .orElse(Stream.empty())
+                    .filter(allocationRange -> allocationRange.getLength() > 0) // filter out 0% allocation ranges
+                    .map(Range::getBucketValue)
+                    .min(Integer::compareTo) // find the minimum bucket value
+                    .flatMap(minActiveBucketValue -> Optional.ofNullable(getTestBucket(minActiveBucketValue)))
+                    .map(minActiveBucket -> new Result(minActiveBucket, null))
```

Comment:
`new Result(minActiveBucket, result.getAllocation()))`? Although I'm not sure the allocation will be used, this is based on the allocation definition.

Reply:
Well I initially set the allocation, but changed my mind because other force group options result in `null` allocations.

Reply:
I see. Sounds good